src/main/java/net/apnic/rdap/ip/IPStatsResourceLocator.java

```diff
@@ -46,7 +46,7 @@ public RDAPAuthority authorityForResource(IpRange range)
      * {@inheritDocs}
      */
     @Override
-    public void putResourceMapping(IpRange resource, RDAPAuthority authority)
+    public synchronized void putResourceMapping(IpRange resource, RDAPAuthority authority)
```

Comment:
Yes there will be synchronization at the hand-off to the aggregating thread, but since this would be abstracted behind the Future interface you'd be fairly well protected from making mistakes. The new aggregation could then be swapped with the last aggregation that is serving the queries via a volatile reference or a HotSwappableTargetSource removing the need to synchronize the reads.